deepinv/models/ram.py

```diff
@@ -0,0 +1,928 @@
+                torch.nn.Conv2d(c_in, c, kernel_size, padding=padding, bias=bias),
+            )
+            setattr(self, f"skipconv{i}", torch.nn.Conv2d(c_in, c, 1, bias=False))
+
+    def forward(self, x):
+
+        if self.skip_in and self.relu_in:
+            x = self.nl_1(self.convin(x)) + self.zero_conv_skip(x)
+        elif self.skip_in and not self.relu_in:
+            x = self.convin(x) + self.zero_conv_skip(x)
+        else:
+            x = self.convin(x)
+
+        for i in range(self.depth - 1):
+            aux = getattr(self, f"conv1{i}")(x)
+            aux = self.nl_2(aux)
+            aux_0 = getattr(self, f"conv2{i}")(aux)
+            aux_1 = getattr(self, f"skipconv{i}")(x)
+            x = aux_0 + aux_1
+
+        return x
+
+
+# --------------------------------------------------------------------------------------
+class AffineConv2d(nn.Conv2d):
+    def __init__(
+        self,
+        in_channels,
+        out_channels,
+        kernel_size,
+        mode="affine",
+        bias=False,
+        stride=1,
+        padding=0,
+        dilation=1,
+        groups=1,
+        padding_mode="circular",
+        blind=True,
+    ):
+        if mode == "affine":  # f(a*x + 1) = a*f(x) + 1
+            bias = False
+        super().__init__(
+            in_channels,
+            out_channels,
+            kernel_size,
+            bias=bias,
+            stride=stride,
+            padding=padding,
+            dilation=dilation,
+            groups=groups,
+            padding_mode=padding_mode,
+        )
+        self.blind = blind
+        self.mode = mode
+
+    def affine(self, w):
+        """returns new kernels that encode affine combinations"""
+        return (
+            w.view(self.out_channels, -1).roll(1, 1).view(w.size())
+            - w
+            + 1 / w[0, ...].numel()
+        )
+
+    def forward(self, x):
+        if self.mode != "affine":
+            return super().forward(x)
+        else:
+            kernel = (
+                self.affine(self.weight)
+                if self.blind
+                else torch.cat(
+                    (self.affine(self.weight[:, :-1, :, :]), self.weight[:, -1:, :, :]),
+                    dim=1,
+                )
+            )
+            padding = tuple(
+                elt for elt in reversed(self.padding) for _ in range(2)
+            )  # used to translate padding arg used by Conv module to the ones used by F.pad
+            padding_mode = (
+                self.padding_mode if self.padding_mode != "zeros" else "constant"
+            )  # used to translate padding_mode arg used by Conv module to the ones used by F.pad
+            return F.conv2d(
+                F.pad(x, padding, mode=padding_mode),
+                kernel,
+                stride=self.stride,
+                dilation=self.dilation,
+                groups=self.groups,
+            )
+
+
+"""
+Functional blocks below
+
+Parts of code borrowed from
```

Comment:
I think for borrowing code it's worth properly mentioning them where appropriate in each function's docstring to make sure code is credited in the right place